src/main_nep/dataset.cu

```diff
@@ -63,6 +63,17 @@ void Dataset::copy_structures(std::vector<Structure>& structures_input, int n1,
   }
 }
 
+void Dataset::find_has_type(Parameters& para)
+{
+  has_type.resize((para.num_types + 1) * Nc, false);
+  for (int n = 0; n < Nc; ++n) {
+    has_type[para.num_types * Nc + n] = true;
+    for (int na = 0; na < structures[n].num_atom; ++na) {
+      has_type[structures[n].type[na] * Nc + n] = true;
+    }
+  }
+}
+
```

Comment:
Does this get the number of species? Are the losses then computed with one loss per species?

Reply:
`has_type` can be understood as a 2D array `has_type[num_types+1][num_structures]`. Suppose there are 3 structures, and the first only containes element `A` (type 0), the second only `B` (type 1), the third only `AB`, then the array will be calculated to be:

```
true  false true
false true true
true true true
```

The meaning of `has_type[t][n]` is "the structure n contains type t". The last column is always `true` because each structure contains "some type", which will contribute to the total loss.

Reply:
Ah I see. 
Can you reiterate the part about  "some type"? Why is it necessary? Does it refer to a part of the loss that is independent of type?

Reply:
that is the original total loss, which is required for e.g. reporing the RMES in loss.out.

Reply:
that is, apart from the original total loss, now I calculate extra `num_types` losses, which are for A element, B element, C element, ...

Reply:
for the loss of A element, I consider system A (pure element), AB, AC, AD, ..., (binary alloys), ABC, ABC, ABD, ACD, ...

Reply:
To sum up, there are now `num_types + 1` loss function, one is the **global** loss that is used to optimize some global parameters and report the RMSE on screen and loos.out, and the other `num_types` are **local** ones that are used to optimize the different subsets of the parameters exclusively related to each element.

Reply:
Makes sense, thanks!